[android] Crash when you change screen orientation

- Open app
- Open menu burger
- Switch screen orientation
- App crashed
```
09-07 17:49:21.363  7047  7047 E AndroidRuntime: FATAL EXCEPTION: main
09-07 17:49:21.363  7047  7047 E AndroidRuntime: Process: app.organicmaps, PID: 7047
09-07 17:49:21.363  7047  7047 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{app.organicmaps/com.mapswithme.maps.MwmActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.mapswithme.util.bottomsheet.MenuBottomSheetFragment: could not find Fragment constructor
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3676)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3813)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5791)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5682)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:71)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2308)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:106)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.os.Looper.loopOnce(Looper.java:201)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:288)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:7898)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
09-07 17:49:21.363  7047  7047 E AndroidRuntime: Caused by: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.mapswithme.util.bottomsheet.MenuBottomSheetFragment: could not find Fragment constructor
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.Fragment.instantiate(Unknown Source:94)09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentContainer.instantiate(Unknown Source:0)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentManager$2.instantiate(Unknown Source:17)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentState.instantiate(Unknown Source:2)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentStateManager.<init>(Unknown Source:13)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentManager.restoreSaveStateInternal(Unknown Source:119)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentManager.attachController(Unknown Source:140)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentController.attachHost(Unknown Source:4)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentActivity.lambda$init$1(Unknown Source:3)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentActivity.a(Unknown Source:0)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.b.onContextAvailable(Unknown Source:2)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.activity.contextaware.ContextAwareHelper.dispatchOnContextAvailable(Unknown Source:20)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.activity.ComponentActivity.onCreate(Unknown Source:7)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.FragmentActivity.onCreate(Unknown Source:0)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at com.mapswithme.maps.base.BaseMwmFragmentActivity.onCreate(Unknown Source:33)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.Activity.performCreate(Activity.java:8290)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.Activity.performCreate(Activity.java:8269)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1384)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3657)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        ... 15 more
09-07 17:49:21.363  7047  7047 E AndroidRuntime: Caused by: java.lang.NoSuchMethodException: com.mapswithme.util.bottomsheet.MenuBottomSheetFragment.<init> []
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at java.lang.Class.getConstructor0(Class.java:2363)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at java.lang.Class.getConstructor(Class.java:1759)
09-07 17:49:21.363  7047  7047 E AndroidRuntime:        at androidx.fragment.app.Fragment.instantiate(Unknown Source:15)09-07 17:49:21.363  7047  7047 E AndroidRuntime:        ... 33 more
```
Pixel 6 - Android 13
version 2022.08.23-5-Google


Problem is fixed on [debug version](https://github.com/organicmaps/organicmaps/actions/runs/3008434506)

But in debug version menu burger is hide when you change screen orientation and you can't expand menu.

**Before (debug version)**
<img src="https://user-images.githubusercontent.com/87148630/188924065-3961eeb7-2838-4f80-8bfb-faf49e03ea46.png" height="500"/>

**After (debug version)**
<img src="https://user-images.githubusercontent.com/87148630/188924326-52bb072f-981f-409a-85a9-cce8b9cdd8d3.png" width="500"/>

cc @arnaudvergnet

--- a/android/src/com/mapswithme/maps/MwmActivity.java
+++ b/android/src/com/mapswithme/maps/MwmActivity.java
@@ -103,6 +103,8 @@ import java.util.ArrayList;
 import java.util.Objects;
 import java.util.Stack;
 
+import static com.mapswithme.maps.widget.placepage.PlacePageButtons.PLACEPAGE_MORE_MENU_ID;
+
 public class MwmActivity extends BaseMwmFragmentActivity
     implements PlacePageActivationListener,
                View.OnTouchListener,
@@ -117,7 +119,9 @@ public class MwmActivity extends BaseMwmFragmentActivity
                PlacePageController.SlideListener,
                AlertDialogCallback, RoutingModeListener,
                AppBackgroundTracker.OnTransitionListener,
-               NoConnectionListener
+               NoConnectionListener,
+               MenuBottomSheetFragment.MenuBottomSheetInterfaceWithHeader,
+               ToggleMapLayerFragment.LayerItemClickListener
 {
   public static final String EXTRA_TASK = "map_task";
   public static final String EXTRA_LAUNCH_BY_DEEP_LINK = "launch_by_deep_link";
@@ -140,6 +144,9 @@ public class MwmActivity extends BaseMwmFragmentActivity
 
   public static final String ERROR_DRIVING_OPTIONS_DIALOG_TAG = "error_driving_options_dialog_tag";
   private static final String ISOLINES_ERROR_DIALOG_TAG = "isolines_dialog_tag";
+
+  private static final String MAIN_MENU_ID = "MAIN_MENU_BOTTOM_SHEET";
+  private static final String LAYERS_MENU_ID = "LAYERS_MENU_BOTTOM_SHEET";
 
   // Map tasks that we run AFTER rendering initialized
   private final Stack<MapTask> mTasks = new Stack<>();
@@ -166,11 +173,6 @@ public class MwmActivity extends BaseMwmFragmentActivity
   private PanelAnimator mPanelAnimator;
   @Nullable
   private OnmapDownloader mOnmapDownloader;
-
-  @NonNull
-  private MenuBottomSheetFragment mLayersBottomSheet;
-  @NonNull
-  private MenuBottomSheetFragment mMainMenuBottomSheet;
 
   @Nullable
   private MapButtonsController mMapButtonsController;
@@ -407,8 +409,6 @@ public class MwmActivity extends BaseMwmFragmentActivity
                      .getViewTreeObserver()
                      .addOnGlobalLayoutListener(new ToolbarLayoutChangeListener());
 
-    initBottomSheets();
-
     boolean isLaunchByDeepLink = getIntent().getBooleanExtra(EXTRA_LAUNCH_BY_DEEP_LINK, false);
     initViews(isLaunchByDeepLink);
 
@@ -457,35 +457,10 @@ public class MwmActivity extends BaseMwmFragmentActivity
     return windowInsets;
   }
 
-  private void initBottomSheets()
-  {
-    ToggleMapLayerFragment toggleMapLayerFragment = new ToggleMapLayerFragment(this::onLayerItemClicked);
-    mLayersBottomSheet = new MenuBottomSheetFragment(toggleMapLayerFragment);
-    mMainMenuBottomSheet = new MenuBottomSheetFragment(getMainMenuItems());
-  }
-
   private int getDownloadMapsCounter()
   {
     UpdateInfo info = MapManager.nativeGetUpdateInfo(null);
     return info == null ? 0 : info.filesCount;
-  }
-
-  private ArrayList<MenuBottomSheetItem> getMainMenuItems()
-  {
-    ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
-    items.add(new MenuBottomSheetItem(R.string.placepage_add_place_button, R.drawable.ic_plus, this::onAddPlaceOptionSelected));
-    items.add(new MenuBottomSheetItem(
-        R.string.download_maps,
-        R.drawable.ic_download,
-        getDownloadMapsCounter(),
-        this::onDownloadMapsOptionSelected
-    ));
-    mDonatesUrl = Config.getDonateUrl();
-    if (!TextUtils.isEmpty(mDonatesUrl))
-      items.add(new MenuBottomSheetItem(R.string.donate, R.drawable.ic_donate, this::onDonateOptionSelected));
-    items.add(new MenuBottomSheetItem(R.string.settings, R.drawable.ic_settings, this::onSettingsOptionSelected));
-    items.add(new MenuBottomSheetItem(R.string.share_my_location, R.drawable.ic_share, this::onShareLocationOptionSelected));
-    return items;
   }
 
   @Override
@@ -702,7 +677,7 @@ public class MwmActivity extends BaseMwmFragmentActivity
         break;
       case menu:
         closeFloatingPanels();
-        showMainMenuBottomSheet();
+        showBottomSheet(MAIN_MENU_ID);
         break;
       case help:
         showHelp();
@@ -711,37 +686,25 @@ public class MwmActivity extends BaseMwmFragmentActivity
   }
 
 
+  private boolean closeBottomSheet(String id)
+  {
+    MenuBottomSheetFragment bottomSheet =
+        (MenuBottomSheetFragment) getSupportFragmentManager().findFragmentByTag(id);
+    if (bottomSheet == null || !bottomSheet.isAdded())
+      return false;
+    bottomSheet.dismiss();
+    return true;
+  }
+
+  private void showBottomSheet(String id)
+  {
+    MenuBottomSheetFragment.newInstance(id).show(getSupportFragmentManager(), id);
+  }
+
   private void toggleMapLayerBottomSheet()
   {
-    if (!closeMapLayerBottomSheet())
-      showMapLayerBottomSheet();
-  }
-
-  private boolean closeMapLayerBottomSheet()
-  {
-    if (!mLayersBottomSheet.isAdded())
-      return false;
-    mLayersBottomSheet.dismiss();
-    return true;
-  }
-
-  private void showMapLayerBottomSheet()
-  {
-    mLayersBottomSheet.show(getSupportFragmentManager(), "layersBottomSheet");
-  }
-
-  private boolean closeMainMenuBottomSheet()
-  {
-    if (!mMainMenuBottomSheet.isAdded())
-      return false;
-    mMainMenuBottomSheet.dismiss();
-    return true;
-  }
-
-  private void showMainMenuBottomSheet()
-  {
-    mMainMenuBottomSheet = new MenuBottomSheetFragment(getMainMenuItems());
-    mMainMenuBottomSheet.show(getSupportFragmentManager(), "mainMenuBottomSheet");
+    if (!closeBottomSheet(LAYERS_MENU_ID))
+      showBottomSheet(LAYERS_MENU_ID);
   }
 
   /**
@@ -835,8 +798,8 @@ public class MwmActivity extends BaseMwmFragmentActivity
 
   public void closeFloatingPanels()
   {
-    closeMainMenuBottomSheet();
-    closeMapLayerBottomSheet();
+    closeBottomSheet(LAYERS_MENU_ID);
+    closeBottomSheet(MAIN_MENU_ID);
     closePlacePage();
   }
 
@@ -1155,7 +1118,7 @@ public class MwmActivity extends BaseMwmFragmentActivity
   public void onBackPressed()
   {
     RoutingController routingController = RoutingController.get();
-    if (!closeMainMenuBottomSheet() && !closeMapLayerBottomSheet() && !collapseNavMenu() &&
+    if (!closeBottomSheet(MAIN_MENU_ID) && !closeBottomSheet(LAYERS_MENU_ID) && !collapseNavMenu() &&
         !closePlacePage() &&!closeSearchToolbar(true, true) &&
         !closeSidePanel() && !closePositionChooser() &&
         !routingController.resetToPlanningStateIfNavigating() && !routingController.cancel())
@@ -1970,9 +1933,46 @@ public class MwmActivity extends BaseMwmFragmentActivity
     shareMyLocation();
   }
 
-  public void onLayerItemClicked(@NonNull Mode mode)
+  @Override
+  public void onLayerItemClick(@NonNull Mode mode)
   {
     closeFloatingPanels();
-    mMapButtonsController.toggleMapLayer(mode);
+    if (mMapButtonsController != null)
+      mMapButtonsController.toggleMapLayer(mode);
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id)
+  {
+    if (id.equals(MAIN_MENU_ID))
+    {
+      ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
+      items.add(new MenuBottomSheetItem(R.string.placepage_add_place_button, R.drawable.ic_plus, this::onAddPlaceOptionSelected));
+      items.add(new MenuBottomSheetItem(
+          R.string.download_maps,
+          R.drawable.ic_download,
+          getDownloadMapsCounter(),
+          this::onDownloadMapsOptionSelected
+      ));
+      mDonatesUrl = Config.getDonateUrl();
+      if (!TextUtils.isEmpty(mDonatesUrl))
+        items.add(new MenuBottomSheetItem(R.string.donate, R.drawable.ic_donate, this::onDonateOptionSelected));
+      items.add(new MenuBottomSheetItem(R.string.settings, R.drawable.ic_settings, this::onSettingsOptionSelected));
+      items.add(new MenuBottomSheetItem(R.string.share_my_location, R.drawable.ic_share, this::onShareLocationOptionSelected));
+      return items;
+    }
+    else if (id.equals(PLACEPAGE_MORE_MENU_ID))
+      return mPlacePageController.getMenuBottomSheetItems();
+    return null;
+  }
+
+  @Override
+  @Nullable
+  public Fragment getMenuBottomSheetFragment(String id)
+  {
+    if (id.equals(LAYERS_MENU_ID))
+      return new ToggleMapLayerFragment();
+    return null;
   }
 }

--- a/android/src/com/mapswithme/maps/bookmarks/BookmarkCategoriesFragment.java
+++ b/android/src/com/mapswithme/maps/bookmarks/BookmarkCategoriesFragment.java
@@ -45,7 +45,9 @@ public class BookmarkCategoriesFragment extends BaseMwmRecyclerFragment<Bookmark
                CategoryListCallback,
                OnItemClickListener<BookmarkCategory>,
                OnItemMoreClickListener<BookmarkCategory>,
-               OnItemLongClickListener<BookmarkCategory>, BookmarkManager.BookmarksSharingListener
+               OnItemLongClickListener<BookmarkCategory>,
+               BookmarkManager.BookmarksSharingListener,
+               MenuBottomSheetFragment.MenuBottomSheetInterface
 
 {
   private static final String TAG = BookmarkCategoriesFragment.class.getSimpleName();
@@ -54,6 +56,8 @@ public class BookmarkCategoriesFragment extends BaseMwmRecyclerFragment<Bookmark
   static final int REQ_CODE_IMPORT_DIRECTORY = 103;
 
   private static final int MAX_CATEGORY_NAME_LENGTH = 60;
+
+  public static final String BOOKMARKS_CATEGORIES_MENU_ID = "BOOKMARKS_CATEGORIES_BOTTOM_SHEET";
 
   @Nullable
   private BookmarkCategory mSelectedCategory;
@@ -152,23 +156,36 @@ public class BookmarkCategoriesFragment extends BaseMwmRecyclerFragment<Bookmark
   protected final void showBottomMenu(@NonNull BookmarkCategory item)
   {
     mSelectedCategory = item;
-    new MenuBottomSheetFragment(item.getName(), getMenuItems(item))
-            .show(getParentFragmentManager(), "bookmarkCategoriesBottomSheet");
+    MenuBottomSheetFragment.newInstance(BOOKMARKS_CATEGORIES_MENU_ID, item.getName())
+            .show(getChildFragmentManager(), BOOKMARKS_CATEGORIES_MENU_ID);
   }
 
-  private ArrayList<MenuBottomSheetItem> getMenuItems(@NonNull BookmarkCategory item)
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id)
   {
     ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
-    items.add(new MenuBottomSheetItem(R.string.list_settings, R.drawable.ic_settings, () -> onSettingsActionSelected(item)));
-    items.add(new MenuBottomSheetItem(
-            item.isVisible() ? R.string.hide : R.string.show,
-            item.isVisible() ? R.drawable.ic_hide : R.drawable.ic_show,
-            () -> onShowActionSelected(item)));
-    items.add(new MenuBottomSheetItem(R.string.export_file, R.drawable.ic_share, () -> onShareActionSelected(item)));
-    // Disallow deleting the last category
-    if (getAdapter().getBookmarkCategories().size() > 1)
-      items.add(new MenuBottomSheetItem(R.string.delete, R.drawable.ic_delete, () -> onDeleteActionSelected(item)));
-
+    if (mSelectedCategory != null)
+    {
+      items.add(new MenuBottomSheetItem(
+          R.string.list_settings,
+          R.drawable.ic_settings,
+          () -> onSettingsActionSelected(mSelectedCategory)));
+      items.add(new MenuBottomSheetItem(
+          mSelectedCategory.isVisible() ? R.string.hide : R.string.show,
+          mSelectedCategory.isVisible() ? R.drawable.ic_hide : R.drawable.ic_show,
+          () -> onShowActionSelected(mSelectedCategory)));
+      items.add(new MenuBottomSheetItem(
+          R.string.export_file,
+          R.drawable.ic_share,
+          () -> onShareActionSelected(mSelectedCategory)));
+      // Disallow deleting the last category
+      if (getAdapter().getBookmarkCategories().size() > 1)
+        items.add(new MenuBottomSheetItem(
+            R.string.delete,
+            R.drawable.ic_delete,
+            () -> onDeleteActionSelected(mSelectedCategory)));
+    }
     return items;
   }
 

--- a/android/src/com/mapswithme/maps/bookmarks/BookmarksListFragment.java
+++ b/android/src/com/mapswithme/maps/bookmarks/BookmarksListFragment.java
@@ -53,13 +53,17 @@ public class BookmarksListFragment extends BaseMwmRecyclerFragment<ConcatAdapter
                BookmarkManager.BookmarksSortingListener,
                BookmarkManager.BookmarksLoadingListener,
                NativeBookmarkSearchListener,
-               ChooseBookmarksSortingTypeFragment.ChooseSortingTypeListener
+               ChooseBookmarksSortingTypeFragment.ChooseSortingTypeListener,
+               MenuBottomSheetFragment.MenuBottomSheetInterface
 {
   public static final String TAG = BookmarksListFragment.class.getSimpleName();
   public static final String EXTRA_CATEGORY = "bookmark_category";
   public static final String EXTRA_BUNDLE = "bookmark_bundle";
   private static final int INDEX_BOOKMARKS_COLLECTION_ADAPTER = 0;
   private static final int INDEX_BOOKMARKS_LIST_ADAPTER = 1;
+  private static final String BOOKMARKS_MENU_ID = "BOOKMARKS_MENU_BOTTOM_SHEET";
+  private static final String TRACK_MENU_ID = "TRACK_MENU_BOTTOM_SHEET";
+  private static final String OPTIONS_MENU_ID = "OPTIONS_MENU_BOTTOM_SHEET";
 
   @SuppressWarnings("NotNullFieldNotInitialized")
   @NonNull
@@ -616,14 +620,14 @@ public class BookmarksListFragment extends BaseMwmRecyclerFragment<ConcatAdapter
 
       case BookmarkListAdapter.TYPE_BOOKMARK:
         final BookmarkInfo bookmark = (BookmarkInfo) adapter.getItem(mSelectedPosition);
-        new MenuBottomSheetFragment(bookmark.getName(), getBookmarkMenuItems())
-                .show(getParentFragmentManager(), "bookmarkBottomSheet");
+        MenuBottomSheetFragment.newInstance(BOOKMARKS_MENU_ID, bookmark.getName())
+                .show(getChildFragmentManager(), BOOKMARKS_MENU_ID);
         break;
 
       case BookmarkListAdapter.TYPE_TRACK:
         final Track track = (Track) adapter.getItem(mSelectedPosition);
-        new MenuBottomSheetFragment(track.getName(), getTrackMenuItems(track))
-            .show(getParentFragmentManager(), "trackBottomSheet");
+        MenuBottomSheetFragment.newInstance(TRACK_MENU_ID, track.getName())
+            .show(getChildFragmentManager(), TRACK_MENU_ID);
         break;
     }
   }
@@ -669,8 +673,8 @@ public class BookmarksListFragment extends BaseMwmRecyclerFragment<ConcatAdapter
 
     if (item.getItemId() == R.id.bookmarks_more)
     {
-      new MenuBottomSheetFragment(mCategoryDataSource.getData().getName(), getOptionsMenuItems())
-              .show(getParentFragmentManager(), "optionsBottomSheet");
+      MenuBottomSheetFragment.newInstance(OPTIONS_MENU_ID, mCategoryDataSource.getData().getName())
+              .show(getChildFragmentManager(), OPTIONS_MENU_ID);
       return true;
     }
 
@@ -811,5 +815,21 @@ public class BookmarksListFragment extends BaseMwmRecyclerFragment<ConcatAdapter
     View loadingPlaceholder = root.findViewById(R.id.placeholder_loading);
     UiUtils.showIf(!isShowLoadingPlaceholder, root, R.id.show_on_map_fab);
     UiUtils.showIf(isShowLoadingPlaceholder, loadingPlaceholder);
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id)
+  {
+    if (id.equals(BOOKMARKS_MENU_ID))
+      return getBookmarkMenuItems();
+    if (id.equals(TRACK_MENU_ID))
+    {
+      final Track track = (Track) getBookmarkListAdapter().getItem(mSelectedPosition);
+      return getTrackMenuItems(track);
+    }
+    if (id.equals(OPTIONS_MENU_ID))
+      return getOptionsMenuItems();
+    return null;
   }
 }

--- a/android/src/com/mapswithme/maps/downloader/DownloaderAdapter.java
+++ b/android/src/com/mapswithme/maps/downloader/DownloaderAdapter.java
@@ -49,6 +49,8 @@ class DownloaderAdapter extends RecyclerView.Adapter<DownloaderAdapter.ViewHolde
 
   private static final int TYPE_COUNTRY = 0;
 
+  private static final String DOWNLOADER_MENU_ID = "DOWNLOADER_BOTTOM_SHEET";
+
   private final RecyclerView mRecycler;
   private final Activity mActivity;
   private final DownloaderFragment mFragment;
@@ -68,6 +70,7 @@ class DownloaderAdapter extends RecyclerView.Adapter<DownloaderAdapter.ViewHolde
   private int mNearMeCount;
 
   private int mListenerSlot;
+  private CountryItem mSelectedItem;
 
   private void onDownloadActionSelected(final CountryItem item, DownloaderAdapter adapter)
   {
@@ -260,6 +263,94 @@ class DownloaderAdapter extends RecyclerView.Adapter<DownloaderAdapter.ViewHolde
     }
   }
 
+  private void showBottomSheet(CountryItem selectedItem)
+  {
+    mSelectedItem = selectedItem;
+    MenuBottomSheetFragment.newInstance(DOWNLOADER_MENU_ID, mSelectedItem.name)
+                           .show(mFragment.getChildFragmentManager(), DOWNLOADER_MENU_ID);
+  }
+
+  public ArrayList<MenuBottomSheetItem> getMenuItems()
+  {
+    ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
+    switch (mSelectedItem.status)
+    {
+      case CountryItem.STATUS_DOWNLOADABLE:
+        items.add(getDownloadMenuItem());
+        break;
+
+      case CountryItem.STATUS_UPDATABLE:
+        items.add(getUpdateMenuItem());
+        // Fallthrough
+
+      case CountryItem.STATUS_DONE:
+        if (!mSelectedItem.isExpandable())
+          items.add(getExploreMenuItem());
+        appendDeleteMenuItem(items);
+        break;
+
+      case CountryItem.STATUS_FAILED:
+        items.add(getCancelMenuItem());
+
+        if (mSelectedItem.present)
+        {
+          appendDeleteMenuItem(items);
+          items.add(getExploreMenuItem());
+        }
+        break;
+
+      case CountryItem.STATUS_PROGRESS:
+      case CountryItem.STATUS_APPLYING:
+      case CountryItem.STATUS_ENQUEUED:
+        items.add(getCancelMenuItem());
+
+        if (mSelectedItem.present)
+          items.add(getExploreMenuItem());
+        break;
+
+      case CountryItem.STATUS_PARTLY:
+        items.add(getDownloadMenuItem());
+        appendDeleteMenuItem(items);
+        break;
+    }
+    return items;
+  }
+
+
+  private MenuBottomSheetItem getDownloadMenuItem()
+  {
+    return new MenuBottomSheetItem(R.string.downloader_download_map, R.drawable.ic_download,
+                                   () -> onDownloadActionSelected(mSelectedItem, DownloaderAdapter.this));
+  }
+
+  private MenuBottomSheetItem getUpdateMenuItem()
+  {
+    return new MenuBottomSheetItem(R.string.downloader_update_map, R.drawable.ic_update,
+                                   () -> onUpdateActionSelected(mSelectedItem, DownloaderAdapter.this));
+  }
+
+  private MenuBottomSheetItem getExploreMenuItem()
+  {
+    return new MenuBottomSheetItem(R.string.zoom_to_country, R.drawable.ic_explore,
+                                   () -> onExploreActionSelected(mSelectedItem, DownloaderAdapter.this));
+  }
+
+  private void appendDeleteMenuItem(ArrayList<MenuBottomSheetItem> items)
+  {
+    // Do not show "Delete" option for World files.
+    // Checking the name is not beautiful, but the simplest way for now ..
+    if (!mSelectedItem.id.startsWith("World")) {
+      items.add(new MenuBottomSheetItem(R.string.delete, R.drawable.ic_delete,
+                                        () -> onDeleteActionSelected(mSelectedItem, DownloaderAdapter.this)));
+    }
+  }
+
+  private MenuBottomSheetItem getCancelMenuItem()
+  {
+    return new MenuBottomSheetItem(R.string.cancel, R.drawable.ic_cancel,
+                                   () -> onCancelActionSelected(mSelectedItem, DownloaderAdapter.this));
+  }
+
   private class ItemViewHolder extends BaseInnerViewHolder<CountryItem>
   {
     private final DownloaderStatusIcon mStatusIcon;
@@ -311,89 +402,7 @@ class DownloaderAdapter extends RecyclerView.Adapter<DownloaderAdapter.ViewHolde
 
     private void processLongClick()
     {
-      ArrayList<MenuBottomSheetItem> items = getMenuItems();
-      new MenuBottomSheetFragment(mItem.name, items)
-          .show(mFragment.getParentFragmentManager(), "downloaderBottomSheet");
-    }
-
-    private MenuBottomSheetItem getDownloadMenuItem()
-    {
-      return new MenuBottomSheetItem(R.string.downloader_download_map, R.drawable.ic_download,
-          () -> onDownloadActionSelected(mItem, DownloaderAdapter.this));
-    }
-
-    private MenuBottomSheetItem getUpdateMenuItem()
-    {
-      return new MenuBottomSheetItem(R.string.downloader_update_map, R.drawable.ic_update,
-          () -> onUpdateActionSelected(mItem, DownloaderAdapter.this));
-    }
-
-    private MenuBottomSheetItem getExploreMenuItem()
-    {
-      return new MenuBottomSheetItem(R.string.zoom_to_country, R.drawable.ic_explore,
-          () -> onExploreActionSelected(mItem, DownloaderAdapter.this));
-    }
-
-    private void appendDeleteMenuItem(ArrayList<MenuBottomSheetItem> items)
-    {
-      // Do not show "Delete" option for World files.
-      // Checking name is not beautiful, but the simplest way for now ..
-      if (!mItem.id.startsWith("World")) {
-        items.add(new MenuBottomSheetItem(R.string.delete, R.drawable.ic_delete,
-                () -> onDeleteActionSelected(mItem, DownloaderAdapter.this)));
-      }
-    }
-
-    private MenuBottomSheetItem getCancelMenuItem()
-    {
-      return new MenuBottomSheetItem(R.string.cancel, R.drawable.ic_cancel,
-          () -> onCancelActionSelected(mItem, DownloaderAdapter.this));
-    }
-
-    private ArrayList<MenuBottomSheetItem> getMenuItems()
-    {
-      ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
-      switch (mItem.status)
-      {
-        case CountryItem.STATUS_DOWNLOADABLE:
-          items.add(getDownloadMenuItem());
-          break;
-
-        case CountryItem.STATUS_UPDATABLE:
-          items.add(getUpdateMenuItem());
-          // Fallthrough
-
-        case CountryItem.STATUS_DONE:
-          if (!mItem.isExpandable())
-            items.add(getExploreMenuItem());
-          appendDeleteMenuItem(items);
-          break;
-
-        case CountryItem.STATUS_FAILED:
-          items.add(getCancelMenuItem());
-
-          if (mItem.present)
-          {
-            appendDeleteMenuItem(items);
-            items.add(getExploreMenuItem());
-          }
-          break;
-
-        case CountryItem.STATUS_PROGRESS:
-        case CountryItem.STATUS_APPLYING:
-        case CountryItem.STATUS_ENQUEUED:
-          items.add(getCancelMenuItem());
-
-          if (mItem.present)
-            items.add(getExploreMenuItem());
-          break;
-
-        case CountryItem.STATUS_PARTLY:
-          items.add(getDownloadMenuItem());
-          appendDeleteMenuItem(items);
-          break;
-      }
-      return items;
+      showBottomSheet(mItem);
     }
 
     ItemViewHolder(View frame)

--- a/android/src/com/mapswithme/maps/downloader/DownloaderFragment.java
+++ b/android/src/com/mapswithme/maps/downloader/DownloaderFragment.java
@@ -16,12 +16,15 @@ import com.mapswithme.maps.base.OnBackPressListener;
 import com.mapswithme.maps.search.NativeMapSearchListener;
 import com.mapswithme.maps.search.SearchEngine;
 import com.mapswithme.maps.widget.PlaceholderView;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetFragment;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class DownloaderFragment extends BaseMwmRecyclerFragment<DownloaderAdapter>
-                             implements OnBackPressListener
+                             implements OnBackPressListener,
+                                        MenuBottomSheetFragment.MenuBottomSheetInterface
 {
   private DownloaderToolbarController mToolbarController;
 
@@ -229,5 +232,12 @@ public class DownloaderFragment extends BaseMwmRecyclerFragment<DownloaderAdapte
       placeholder.setContent(R.string.search_not_found, R.string.search_not_found_query);
     else
       placeholder.setContent(R.string.downloader_no_downloaded_maps_title, R.string.downloader_no_downloaded_maps_message);
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id)
+  {
+    return mAdapter != null ? mAdapter.getMenuItems() : null;
   }
 }

--- a/android/src/com/mapswithme/maps/maplayer/ToggleMapLayerFragment.java
+++ b/android/src/com/mapswithme/maps/maplayer/ToggleMapLayerFragment.java
@@ -23,21 +23,20 @@ import java.util.List;
 
 public class ToggleMapLayerFragment extends Fragment
 {
-  @NonNull
-  private final LayerItemClickListener mLayerItemClickListener;
+  @Nullable
+  private LayerItemClickListener mLayerItemClickListener;
   @Nullable
   private LayersAdapter mAdapter;
-
-  public ToggleMapLayerFragment(@NonNull LayerItemClickListener layerItemClickListener)
-  {
-    mLayerItemClickListener = layerItemClickListener;
-  }
 
   @Nullable
   @Override
   public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
   {
     View mRoot = inflater.inflate(R.layout.fragment_toggle_map_layer, container, false);
+
+    if (getActivity() instanceof LayerItemClickListener)
+      mLayerItemClickListener = ((LayerItemClickListener) getActivity());
+
     initRecycler(mRoot);
     return mRoot;
   }
@@ -73,11 +72,12 @@ public class ToggleMapLayerFragment extends Fragment
     mAdapter.notifyDataSetChanged();
     if (IsolinesManager.from(context).shouldShowNotification())
       Utils.showSnackbar(context, v.getRootView(), R.string.isolines_toast_zooms_1_10);
-    mLayerItemClickListener.onClick(mode);
+    if (mLayerItemClickListener != null)
+      mLayerItemClickListener.onLayerItemClick(mode);
   }
 
   public interface LayerItemClickListener
   {
-    void onClick(@NonNull Mode mode);
+    void onLayerItemClick(@NonNull Mode mode);
   }
 }

--- a/android/src/com/mapswithme/maps/widget/placepage/PlacePageButtons.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/PlacePageButtons.java
@@ -9,6 +9,7 @@ import android.widget.TextView;
 
 import androidx.annotation.DrawableRes;
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import androidx.annotation.StringRes;
 
 import com.mapswithme.maps.MwmApplication;
@@ -25,6 +26,7 @@ import java.util.List;
 
 public final class PlacePageButtons
 {
+  public static final String PLACEPAGE_MORE_MENU_ID = "PLACEPAGE_MORE_MENU_BOTTOM_SHEET";
   private final int mMaxButtons;
 
   private final PlacePageView mPlacePage;
@@ -32,6 +34,7 @@ public final class PlacePageButtons
   private final ItemListener mItemListener;
 
   private List<PlacePageButton> mPrevItems;
+  private List<PlacePageButton>  mMoreItems;
 
   interface PlacePageButton
   {
@@ -308,22 +311,25 @@ public final class PlacePageButtons
     }
   }
 
-  private ArrayList<MenuBottomSheetItem> getMenuItems(final List<PlacePageButton> buttons)
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems()
   {
+    if (mMoreItems == null)
+      return null;
     ArrayList<MenuBottomSheetItem> items = new ArrayList<>();
-    for (int i = mMaxButtons; i < buttons.size(); i++)
+    for (int i = mMaxButtons; i < mMoreItems.size(); i++)
     {
-      PlacePageButton bsItem = buttons.get(i);
+      final PlacePageButton bsItem = mMoreItems.get(i);
       int iconRes = bsItem.getIcon().getEnabledStateResId(mPlacePage.getContext());
       items.add(new MenuBottomSheetItem(bsItem.getTitle(), iconRes, () -> mItemListener.onItemClick(bsItem)));
     }
     return items;
   }
 
-  private void showPopup(final List<PlacePageButton> buttons)
+  private void showPopup()
   {
-    new MenuBottomSheetFragment(getMenuItems(buttons))
-        .show(mPlacePage.getActivity().getSupportFragmentManager(), "moreBottomSheet");
+    MenuBottomSheetFragment.newInstance(PLACEPAGE_MORE_MENU_ID)
+        .show(mPlacePage.getActivity().getSupportFragmentManager(), PLACEPAGE_MORE_MENU_ID);
   }
 
   private View createButton(@NonNull final List<PlacePageButton> items,
@@ -339,7 +345,9 @@ public final class PlacePageButtons
     title.setText(current.getTitle());
     icon.setImageResource(current.getIcon().getEnabledStateResId(context));
     mItemListener.onPrepareVisibleView(current, parent, icon, title);
-    parent.setOnClickListener(new ShowPopupClickListener(current, items));
+    if (current == Item.MORE)
+      mMoreItems = items;
+    parent.setOnClickListener(new ShowPopupClickListener(current));
     return parent;
   }
 
@@ -361,21 +369,17 @@ public final class PlacePageButtons
   {
     @NonNull
     private final PlacePageButton mCurrent;
-    @NonNull
-    private final List<PlacePageButton> mItems;
 
-    public ShowPopupClickListener(@NonNull PlacePageButton current,
-                                  @NonNull List<PlacePageButton> items)
+    public ShowPopupClickListener(@NonNull PlacePageButton current)
     {
       mCurrent = current;
-      mItems = items;
     }
 
     @Override
     public void onClick(View v)
     {
       if (mCurrent == Item.MORE)
-        showPopup(mItems);
+        showPopup();
       else
         mItemListener.onItemClick(mCurrent);
     }

--- a/android/src/com/mapswithme/maps/widget/placepage/PlacePageController.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/PlacePageController.java
@@ -6,9 +6,13 @@ import android.os.Bundle;
 
 import androidx.annotation.NonNull;
 
+import androidx.annotation.Nullable;
 import com.mapswithme.maps.base.Initializable;
 import com.mapswithme.maps.base.Savable;
 import com.mapswithme.maps.base.Supportable;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
+
+import java.util.ArrayList;
 
 public interface PlacePageController extends Initializable<Activity>,
                                              Savable<Bundle>,
@@ -19,6 +23,8 @@ public interface PlacePageController extends Initializable<Activity>,
   void close(boolean deactivateMapSelection);
   boolean isClosed();
   int getPlacePageWidth();
+  @Nullable
+  ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems();
 
   interface SlideListener
   {

--- a/android/src/com/mapswithme/maps/widget/placepage/PlacePageControllerComposite.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/PlacePageControllerComposite.java
@@ -5,6 +5,7 @@ import android.os.Bundle;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -32,6 +33,13 @@ class PlacePageControllerComposite implements PlacePageController
   public int getPlacePageWidth()
   {
     return mActiveController.getPlacePageWidth();
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems()
+  {
+    return mActiveController.getMenuBottomSheetItems();
   }
 
   @Override

--- a/android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java
@@ -65,6 +65,7 @@ import com.mapswithme.util.StringUtils;
 import com.mapswithme.util.ThemeUtils;
 import com.mapswithme.util.UiUtils;
 import com.mapswithme.util.Utils;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
 import com.mapswithme.util.concurrency.UiThread;
 import com.mapswithme.util.log.Logger;
 
@@ -1633,6 +1634,12 @@ public class PlacePageView extends NestedScrollViewClickFixed
   int getPreviewHeight()
   {
     return mPreview.getHeight();
+  }
+
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems()
+  {
+    return mButtons.getMenuBottomSheetItems();
   }
 
   private class EditBookmarkClickListener implements OnClickListener

--- a/android/src/com/mapswithme/maps/widget/placepage/RichPlacePageController.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/RichPlacePageController.java
@@ -20,8 +20,10 @@ import com.mapswithme.maps.bookmarks.data.MapObject;
 import com.mapswithme.maps.location.LocationHelper;
 import com.mapswithme.maps.location.LocationListener;
 import com.mapswithme.util.UiUtils;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
 import com.mapswithme.util.log.Logger;
 
+import java.util.ArrayList;
 import java.util.Objects;
 
 public class RichPlacePageController implements PlacePageController, LocationListener,
@@ -142,6 +144,13 @@ public class RichPlacePageController implements PlacePageController, LocationLis
   public int getPlacePageWidth()
   {
     return mPlacePage.getWidth();
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems()
+  {
+    return mPlacePage.getMenuBottomSheetItems();
   }
 
   @Override

--- a/android/src/com/mapswithme/maps/widget/placepage/SimplePlacePageController.java
+++ b/android/src/com/mapswithme/maps/widget/placepage/SimplePlacePageController.java
@@ -16,7 +16,9 @@ import com.google.android.material.bottomsheet.BottomSheetBehavior;
 import com.mapswithme.maps.Framework;
 import com.mapswithme.maps.R;
 import com.mapswithme.util.UiUtils;
+import com.mapswithme.util.bottomsheet.MenuBottomSheetItem;
 
+import java.util.ArrayList;
 import java.util.Objects;
 
 public class SimplePlacePageController implements PlacePageController
@@ -113,6 +115,13 @@ public class SimplePlacePageController implements PlacePageController
   public int getPlacePageWidth()
   {
     return mSheet.getWidth();
+  }
+
+  @Override
+  @Nullable
+  public ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems()
+  {
+    return null;
   }
 
   @Override

--- a/android/src/com/mapswithme/util/bottomsheet/MenuBottomSheetFragment.java
+++ b/android/src/com/mapswithme/util/bottomsheet/MenuBottomSheetFragment.java
@@ -1,5 +1,6 @@
 package com.mapswithme.util.bottomsheet;
 
+import android.app.Activity;
 import android.os.Bundle;
 import android.view.LayoutInflater;
 import android.view.View;
@@ -11,64 +12,38 @@ import androidx.annotation.Nullable;
 import androidx.fragment.app.Fragment;
 import androidx.recyclerview.widget.LinearLayoutManager;
 import androidx.recyclerview.widget.RecyclerView;
-
 import com.google.android.material.bottomsheet.BottomSheetBehavior;
 import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
 import com.mapswithme.maps.R;
-import com.mapswithme.util.ThemeUtils;
+import com.mapswithme.util.UiUtils;
 
 import java.util.ArrayList;
 
 public class MenuBottomSheetFragment extends BottomSheetDialogFragment
 {
+
   @Nullable
-  private final String title;
+  private ArrayList<MenuBottomSheetItem> mMenuBottomSheetItems;
   @Nullable
-  private final Fragment headerFragment;
-  private final ArrayList<MenuBottomSheetItem> menuBottomSheetItems;
+  private Fragment mHeaderFragment;
 
-  public MenuBottomSheetFragment(@NonNull Fragment headerFragment, ArrayList<MenuBottomSheetItem> menuBottomSheetItems)
+  public static MenuBottomSheetFragment newInstance(String id)
   {
-    this.title = null;
-    this.headerFragment = headerFragment;
-    this.menuBottomSheetItems = menuBottomSheetItems;
+    Bundle args = new Bundle();
+    args.putString("id", id);
+    MenuBottomSheetFragment f = new MenuBottomSheetFragment();
+    f.setArguments(args);
+    return f;
   }
 
-  public MenuBottomSheetFragment(@NonNull String title, ArrayList<MenuBottomSheetItem> menuBottomSheetItems)
+  public static MenuBottomSheetFragment newInstance(String id, String title)
   {
-    this.title = title;
-    this.headerFragment = null;
-    this.menuBottomSheetItems = menuBottomSheetItems;
-  }
-
-  public MenuBottomSheetFragment(ArrayList<MenuBottomSheetItem> menuBottomSheetItems)
-  {
-    this.title = null;
-    this.headerFragment = null;
-    this.menuBottomSheetItems = menuBottomSheetItems;
-  }
-
-  public MenuBottomSheetFragment(@NonNull Fragment headerFragment)
-  {
-    this.title = null;
-    this.headerFragment = headerFragment;
-    this.menuBottomSheetItems = new ArrayList<>();
-  }
-
-  // https://stackoverflow.com/questions/51831053/could-not-find-fragment-constructor
-  public MenuBottomSheetFragment()
-  {
-    this.title = null;
-    this.headerFragment = null;
-    /// @todo Make an empty list as in ctor above. Can we leave it null and don't create MenuAdapter?
-    /// What is the reason of MenuAdapter with empty items list?
-    this.menuBottomSheetItems = new ArrayList<>();
-  }
-
-  @Override
-  public void onCreate(@Nullable Bundle savedInstanceState)
-  {
-    super.onCreate(savedInstanceState);
+    Bundle args = new Bundle();
+    args.putString("id", id);
+    args.putString("title", title);
+    MenuBottomSheetFragment f = new MenuBottomSheetFragment();
+    f.setArguments(args);
+    return f;
   }
 
   @Nullable
@@ -93,19 +68,93 @@ public class MenuBottomSheetFragment extends BottomSheetDialogFragment
   public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState)
   {
     super.onViewCreated(view, savedInstanceState);
+    attachToNearestContext();
     TextView titleView = view.findViewById(R.id.bottomSheetTitle);
     RecyclerView recyclerView = view.findViewById(R.id.bottomSheetMenuContainer);
-    if (title != null)
+    if (getArguments() != null)
     {
-      titleView.setVisibility(View.VISIBLE);
-      titleView.setText(title);
-    } else
+      String title = getArguments().getString("title");
+      if (title != null && title.length() > 0)
+      {
+        titleView.setVisibility(View.VISIBLE);
+        titleView.setText(title);
+      }
+      else
+        titleView.setVisibility(View.GONE);
+    }
+    else
       titleView.setVisibility(View.GONE);
 
-    MenuAdapter menuAdapter = new MenuAdapter(menuBottomSheetItems, this::dismiss);
-    recyclerView.setAdapter(menuAdapter);
-    recyclerView.setLayoutManager(new LinearLayoutManager(requireActivity()));
-    if (headerFragment != null)
-      getChildFragmentManager().beginTransaction().add(R.id.bottom_sheet_menu_header, headerFragment).commit();
+    if (mMenuBottomSheetItems != null)
+    {
+      MenuAdapter menuAdapter = new MenuAdapter(mMenuBottomSheetItems, this::dismiss);
+      recyclerView.setAdapter(menuAdapter);
+      recyclerView.setLayoutManager(new LinearLayoutManager(requireActivity()));
+    }
+    if (mHeaderFragment != null)
+      getChildFragmentManager().beginTransaction().add(R.id.bottom_sheet_menu_header, mHeaderFragment).commit();
+
+    // If there is nothing to show, hide the sheet
+    if (!UiUtils.isVisible(titleView) && mMenuBottomSheetItems == null && mHeaderFragment == null)
+      dismiss();
   }
+
+  private void attachToNearestContext()
+  {
+    // Try to attach to the parent fragment if any
+    // In other cases, attach to the activity
+    MenuBottomSheetInterface bottomSheetInterface = null;
+    MenuBottomSheetInterfaceWithHeader bottomSheetInterfaceWithHeader = null;
+
+    Fragment parentFragment = getParentFragment();
+    if (parentFragment instanceof MenuBottomSheetInterfaceWithHeader)
+      bottomSheetInterfaceWithHeader = (MenuBottomSheetInterfaceWithHeader) parentFragment;
+    else if (parentFragment instanceof MenuBottomSheetInterface)
+      bottomSheetInterface = (MenuBottomSheetInterface) parentFragment;
+    else
+    {
+      Activity parentActivity = getActivity();
+      if (parentActivity instanceof MenuBottomSheetInterfaceWithHeader)
+        bottomSheetInterfaceWithHeader  = (MenuBottomSheetInterfaceWithHeader) parentActivity;
+      else if (parentActivity instanceof MenuBottomSheetInterface)
+        bottomSheetInterface  = (MenuBottomSheetInterface) parentActivity;
+    }
+
+    if (bottomSheetInterface != null)
+    {
+      if (getArguments() != null)
+      {
+        String id = getArguments().getString("id");
+        if (id != null && id.length() > 0)
+          mMenuBottomSheetItems = bottomSheetInterface.getMenuBottomSheetItems(id);
+      }
+    }
+    else if (bottomSheetInterfaceWithHeader != null)
+    {
+      if (getArguments() != null)
+      {
+        String id = getArguments().getString("id");
+        if (id != null && id.length() > 0)
+        {
+          mMenuBottomSheetItems = bottomSheetInterfaceWithHeader.getMenuBottomSheetItems(id);
+          mHeaderFragment = bottomSheetInterfaceWithHeader.getMenuBottomSheetFragment(id);
+        }
+      }
+    }
+  }
+
+  public interface MenuBottomSheetInterfaceWithHeader
+  {
+    @Nullable
+    Fragment getMenuBottomSheetFragment(String id);
+    @Nullable
+    ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id);
+  }
+
+  public interface MenuBottomSheetInterface
+  {
+    @Nullable
+    ArrayList<MenuBottomSheetItem> getMenuBottomSheetItems(String id);
+  }
+
 }

